protostar/utils/data_transformer_test.py

```diff
@@ -0,0 +1,229 @@
+        "stateMutability": "view",
+        "type": "function",
+    },
+    {
+        "inputs": [
+            {"name": "arg", "type": "Test"},
+            {"name": "felt_arg", "type": "felt"},
+            {"name": "uint256_arg", "type": "Uint256"},
+            {"name": "felt_list_arg_len", "type": "felt"},
+            {"name": "felt_list_arg", "type": "felt*"},
+            {"name": "inst_list_arg_len", "type": "felt"},
+            {"name": "inst_list_arg", "type": "InnerStruct*"},
+        ],
+        "name": "constructor",
+        "outputs": [],
+        "type": "constructor",
+    },
+    {
+        "data": [
+            {"name": "arg", "type": "Test"},
+            {"name": "felt_arg", "type": "felt"},
+            {"name": "uint256_arg", "type": "Uint256"},
+            {"name": "felt_list_arg_len", "type": "felt"},
+            {"name": "felt_list_arg", "type": "felt*"},
+            {"name": "inst_list_arg_len", "type": "felt"},
+            {"name": "inst_list_arg", "type": "InnerStruct*"},
+        ],
+        "name": "constructor_event",
+        "type": "event",
+        "from_address": "0",
+    },
+    {
+        "data": [{"name": "input", "type": "(felt, felt)"}],
+        "name": "tuple_event",
+        "type": "event",
+        "from_address": "0",
+    },
+]
+
+VALID_CONSTRUCTOR_PYTHON_INPUT = {
+    "arg": {"felt_field": 321, "inner_field": {"a": 0xA, "b": 0xB, "c": 0xC}},
+    "felt_arg": 123,
+    "uint256_arg": {"high": 101, "low": 102},
+    #    "felt_list_arg_len": 3,
+    "felt_list_arg": [1, 2, 3],
+    #    "inst_list_arg_len": 2,
+    "inst_list_arg": [{"a": 4, "b": 5, "c": 6}, {"a": 7, "b": 8, "c": 9}],
+}
+
+VALID_CONSTRUCTOR_CAIRO_INPUT = [
+    0xFE17,
+    0xA,
+    0xB,
+    0xC,
+    0xFE17_2,
+    101,
+    102,
+    3,
+    0x1,
+    0x2,
+    0x3,
+    2,
+    0xA,
+    0xB,
+    0xC,
+    0xA2,
+    0xB2,
+    0xC2,
+]
+
+
+@pytest.mark.parametrize(
+    "name,mode,data",
+    [
+        ["constructor", "inputs", {**VALID_CONSTRUCTOR_PYTHON_INPUT, "evil_arg": 42}],
+        ["increase_balance", "outputs", {"???": 0xBAD_C0DE}],
+        [
+            "get_balance",
+            "outputs",
+            {"res": "String definitely longer than 31 characters"},
+        ],
+        ["get_balance", "outputs", {"res": Exception("Definitely not a valid output")}],
+    ],
+)
+def test_from_python_fail(
+    name: str, mode: Literal["inputs", "outputs"], data: PythonData
+):
+    with pytest.raises(DataTransformerException):
+        from_python_transformer(TEST_ABI, name, mode)(data)
+
+
+@pytest.mark.parametrize(
+    "name,mode,data",
+    [
+        ["constructor", "inputs", VALID_CONSTRUCTOR_PYTHON_INPUT],
+        ["increase_balance", "outputs", {}],
+        ["get_balance", "outputs", {"res": 0x42}],
+    ],
+)
+def test_from_python_pass(
+    name: str, mode: Literal["inputs", "outputs"], data: PythonData
+):
+    from_python_transformer(TEST_ABI, name, mode)(data)
```

Comment:
You don't check the output of the transformer

Consider:
```python
cairo_input = from_python(python_input)
assert_cairo_input_is_list(cairo_input)
python_input_2 = to_python(cairo_input)
assert python_input == python_input_2
```

Reply:
done